File: Trie/Python/Remove Sub-Folders from the Filesystem.py
# Problem link: https://leetcode.com/problems/remove-sub-folders-from-the-filesystem/description
# Time: O(N * L)
# Space: O(N * L)
class TrieNode:
    def __init__(self):
        self.store = {}
        self.isEnd = False

class Trie:

    def __init__(self):
        self.root = TrieNode()
    
    def add(self, folder):
        node = self.root

        for fold in folder:
            if fold not in node.store:
                node.store[fold] = TrieNode()
            
            node = node.store[fold]
        
        node.isEnd = True
    
    def search(self, folder):
        node = self.root

        for fold in folder:
            if node.isEnd:
                return False
            
            node = node.store[fold]

        return True


class Solution:
    def removeSubfolders(self, folder: List[str]) -> List[str]:
        trie = Trie()

        for fold in folder:
            trie.add(fold.split('/'))
        
        result = []

        for fold in folder:
            if trie.search(fold.split('/')):
                result.append(fold)
            
        return result

File: Trie/Java/Remove Sub-Folders from the Filesystem.java
// Problem link: https://leetcode.com/problems/remove-sub-folders-from-the-filesystem/description/
// Time: O(n * l)
// Space: O(n * l)
class TrieNode {
    HashMap<String, TrieNode> store;
    boolean isEnd;

    public TrieNode() {
        this.store = new HashMap<>();
        this.isEnd = false;
    }
}

class Trie {
    TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }

    public void add(String[] folder) {
        TrieNode node = this.root;

        for(String fold : folder) {
            if(!node.store.containsKey(fold)) {
                node.store.put(fold, new TrieNode());
            }
            node = node.store.get(fold);
        }

        node.isEnd = true;
    }

    public boolean search(String[] folder) {
        TrieNode node = this.root;

        for(String fold : folder) {
            if(node.isEnd) {
                return false;
            }
            node = node.store.get(fold);
        }

        return true;
    }
}

class Solution {
    public List<String> removeSubfolders(String[] folder) {
        List<String> result = new ArrayList<>();
        Trie trie = new Trie();

        for(String fold : folder) {
            trie.add(fold.split("/"));
        }

        for(String fold : folder) {
            if(trie.search(fold.split("/"))) {
                result.add(fold);
            }
        }

        return result;
    }
}
